### PR TITLE
Improve OSX setup flow based on experience

### PR DESCRIPTION
**Pull request by jdlrobson**
## Description

Changes:
* The code inside run.sh is pulled out into a script file setup-tmp-directories.sh as it needs to be run within the `make test` command as well
* New requirements files are added for Mac OS users.

## Test Plan

- [ ] Confirm that following the README instructions allows to run `make test`, `./run.sh`, and connecting to a development sever successfully on MacOS X

### CODE REVIEW

File: README.md
Comment (gonzalo-bulnes) [reply README.md:32]: I was checking what the best place would be for this information, and I realized that it is [already in the README](https://github.com/freedomofpress/securedrop-client/tree/0.8.1#set-up-the-securedrop-client).

Comment (gonzalo-bulnes) [inline README.md:32]:
```suggestion
```

**Review by gonzalo-bulnes — COMMENTED** [review]
Thank for the PR @jdlrobson! I've listed a few changes after review : )

**Comment by jdlrobson** [comment]
(Sorry for the delay on this one, it's very much on my radar I just haven't had the time yet - aiming to get back to this before end of week!)

File: requirements/dev-macos-requirements.txt
Comment (jdlrobson) [reply requirements/dev-macos-requirements.txt]: I've regenerated it in Python 3.9 rather than 3.10. Is that okay?

File: README.md
Comment (jdlrobson) [reply README.md]: Done!

File: requirements/dev-macos-requirements.txt
Comment (gonzalo-bulnes) [reply requirements/dev-macos-requirements.txt]: Yes, this look good @jdlrobson ! 👍

**Comment by gonzalo-bulnes** [comment]
Take your time @jdlrobson, your work is appreciated!


One last thing: please remove the changes to `requirements/dev-bullseye-requirements.txt` from the change set. On one hand there is no reason why adding support for MacOS would change the Bullseye requirements, and on the other, the additions are specific to MacOS. Removing them should turn make the automated test suite successful.

----

About _why_ those changes appeared when you re-generated the requirements files: you did nothing wrong. We could make separate make targets for each platform, but have so far preferred not to multiply the number of targets. When re-generating files, platform-specific requirements always sneak in, so the simplicity of using a single unified make target comes with the caveat that you need to ignore any changes not relevant to the platform you're generating requirements for. There are compromises either way 🙂

**Review by gonzalo-bulnes — APPROVED** [review]
I've got a few things to fix before this can be merged (see CI report) but the change set here is ready to go. 🎉

**Comment by gonzalo-bulnes** [comment]
@jdlrobson Can I ask you to re-generate the `requirements/dev-macos-requirements.txt` file one more time? We had up upgrade a dependency today and I'd rather make sure the file is properly up-to-date at least when merging. You can rebase onto the latest `main` in order to pick up the changes, and this time the CI pipeline will be green. :green_apple: 

(We'll also have to decide how to maintain the MacOS requirements up-to-date going forward, but that's a distinct question.)

**Comment by gonzalo-bulnes** [comment]
Small follow up on future updates to the MacOS development requirement files: @cfm can apply them if needed! :slightly_smiling_face:

**Comment by gonzalo-bulnes** [comment]
Oh, I think we need to exclude the MacOS requirements file from [**check-testing-requirements**](https://app.circleci.com/pipelines/github/freedomofpress/securedrop-client/2889/workflows/61a3fd89-3368-4fd5-a5ed-bddc7be98500/jobs/15407?invite=true#step-103-4346) because we're not generating the file on MAcOS in CI, the comparison makes no sense.

Can you proof-read my thoughts @cfm? If that makes sense, I can add that commit.

**Comment by gonzalo-bulnes** [comment]
Regarding the failure of [**check-python-security-{bullseye,bookworm}**](https://app.circleci.com/pipelines/github/freedomofpress/securedrop-client/2889/workflows/61a3fd89-3368-4fd5-a5ed-bddc7be98500/jobs/15405): I think everything is said in the name, the MacOS requirements file has no business being checked there either because we don't use it to build any production artifacts.

@cfm Please check if I'm missing something... I can't think why we would that check to be performed. :thought_balloon:

**Comment by gonzalo-bulnes** [comment]
@jdlrobson We're doing some adjustments to to automated checks that I didn't think of before seeing the failing CI builds. I think @cfm and I have the necessary context to make the decisions and implement them, but I'm happy to explain if anything doesn't quite make sense to you :slightly_smiling_face:

**Comment by cfm** [comment]
After a little more digging, I understand why these failures are surprising us here.  In #1410 we (oops; I!) sought

> a compromise between (a) having to maintain a full `dev-mac-requirements.txt` (à la #1335) and (b) not providing *any* support for development on macOS (à la #1268).  In effect, by installing dependencies without `--require-hashes`, we hereby offer macOS support on a best-effort basis, to be used by contributors at their own risk.  (https://github.com/freedomofpress/securedrop-client/pull/1410#issue-1120137051)

Relevant for us here, this approach gave us a `dev-mac-requirements.in` that could be manually `pip install`ed, *without* generating a `dev-mac-requirements.txt` that would be subject to `check-python-security` and `check-testing-requirements`.

There are two ways we can proceed here:

1. keep this new `dev-macos-requirements.{in,txt}`, modify these checks to exclude it, and  delete the old `dev-mac-requirements.in`; or
2. delete `dev-macos-requirements.txt` and apply the changes proposed here to either one of `dev-{mac,macos}-requirements.in`, as last done in #1410.

I have a weak preference for (2), but I'll defer to you, @gonzalo-bulnes, as reviewer.  Happy to elaborate and/or assist further as needed.

**Comment by gonzalo-bulnes** [comment]
:bulb: Does this mean we had MacOS support all along an missed the actual [instructions to set up a developer environment on MacOS](https://github.com/freedomofpress/securedrop-client/tree/0.8.1#developer-environment-on-macos)? :no_mouth:

**Comment by gonzalo-bulnes** [comment]
On the options you describe @cfm, I've a slight preference for 1. because a `requirement.txt` file provides stability to the build environment, and at least makes clear what's out-of-date or different to the canonical (currently Bullseye) environment.

That being said, I'm not going to be the user of this, and I know that, unlike me, you've thought about this before!

**Comment by cfm** [comment]
@gonzalo-bulnes in <https://github.com/freedomofpress/securedrop-client/pull/1592#issuecomment-1331517332>:

> Either way, I'd suggest keeping the `requirements/dev-macos-requirements.in` that @jdlrobson added here, since it is sufficient and looks simpler to the existing `requirements/dev-mac-requirements.in`. I think the differences with the canonical requirements are meaningful, and it makes sense to me to include `dev-bullseye-requirements.in` as a starting point.

The existing `dev-mac-requirements.in` already starts from `dev-bullseye-requirements.in`.  :-)

https://github.com/freedomofpress/securedrop-client/blob/f55606fc16d52943dbafc3905a87f53ca5e34889/requirements/dev-mac-requirements.in#L1

The `dev-macos-requirements.in` proposed here is simpler, but on further review I see it's not actually used:

https://github.com/freedomofpress/securedrop-client/blob/0c3f15a6a4194f591378d009230d5cb5549e63ce/Makefile#L27-L30

However, I can confirm that `make venv-mac` against this `dev-macos-requirements.in` does still succeed.

> On the options you describe @cfm, I've a slight preference for 1. because a `requirement.txt` file provides stability to the build environment, and at least makes clear what's out-of-date or different from the canonical (currently Bullseye) environment.

I respect that preference, _and_ doing it this way (_without_ a pinned `requirements.txt` on macOS) was an explicit decision by @creviera in <https://github.com/freedomofpress/securedrop-client/pull/1396#issuecomment-1026276637> to ease exactly this maintenance burden.  :-)

The compromise I'd suggest would be to rebase along the lines of—

```sh-session
$ git mv requirements/dev-macos-requirements.in requirements/dev-mac-requirements.in  # still used in <https://github.com/freedomofpress/securedrop-client/blob/0c3f15a6a4194f591378d009230d5cb5549e63ce/Makefile#L30>!
$ git rm requirements/dev-macos-requirements.txt  # no longer needed
```

—and keep the contents of this pull request _except_ the `Makefile` additions to re/genenerate `dev-macos-requirements.txt`.  And then I would be happy to document this weird exception!  :-)

**Comment by gonzalo-bulnes** [comment]
Sounds like a plan @cfm ! 👍

**Comment by cfm** [comment]
Rebased from `main` and simplified per <https://github.com/freedomofpress/securedrop-client/pull/1592#issuecomment-1331543254>.  `make venv-mac && ./run.sh` succeeds as expected on my Intel Mac.  @gonzalo-bulnes, I'll merge this tomorrow assuming you don't have any objections on visual review.  :-)

The failure in `functional-test-bullseye`'s `tests/functional/test_offline_send_reply.py::test_offline_send_reply_to_source` is unrelated (occurs starting with last night's `nightly` job), and I'm looking into it now.

**Comment by gonzalo-bulnes** [comment]
Don't worry about the functional test failure @cfm, I restarted a CI run ("from failed") amd it will become greem eventually. It's a flake.

**Review by gonzalo-bulnes — APPROVED** [review]
Looks good @cfm! It is quite nice that the MacOS-specfic requirements input file is getting smaller.

**Review by cfm — APPROVED** [review]
Approving with my changes in <https://github.com/freedomofpress/securedrop-client/pull/1592#issuecomment-1360256593> and @gonzalo-bulnes's confirmation of the flaky test in <https://github.com/freedomofpress/securedrop-client/pull/1592#pullrequestreview-1225197341>.  Thanks for this contribution, @jdlrobson!

**Comment by gonzalo-bulnes** [comment]
Thanks @cfm and thanks @jdlrobson ! :rocket:

**Pull request merged**
